TariikMVP/FEATURE/ProjectsTable

TarikVelic opened the pull request:
--------------------------------------------------------------------
 MAIN
 * Custom Table creation --------------------------------------------------------------------
 OTHER
 * <other msg> --------------------------------------------------------------------

TarikVelic committed:
TariikMVP/FEATURE/ProjectsTable

--------------------------------------------------------------------
 MAIN
 * Custom Table creation
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented:
*components_(component folder in src project folder)
                         |
                          * Login_(main specific component folder)
                                         |
                                         View.js. (main file of component)
                                         index.js (connect Container and View)
                                         Container.js ( file for mapStateToProps...)
                                         *modules_(everything regarding state management goes in this folder and it is just for this 
                                                            component. Of course you can reuse in other modules file for example action)
                                                            |
                                                            reducer.js
                                                            saga.js
                                                            actions.js
                                                            selectors.js
                                                            types.js
                                         *components_
                                                                  |
                                                                  LoginForm.js
                                                                  columns.js
                                                                  UserItem.js
                                                                   (all other components that are included in Login for example itemCard that will be 
                                                                      mapped)

becarselman commented:
<img width="327" alt="image" src="https://github.com/becarselman/ant-colony-pharaoh-frontend/assets/102960105/9d243af7-16d3-4ae0-9488-d656418921fb">
* In src folder we have some other folders for example fonts, images, utils and components
<img width="155" alt="image" src="https://github.com/becarselman/ant-colony-pharaoh-frontend/assets/102960105/a5b6a5c6-7fb8-4d58-a9c3-b0d46a0ff517">
*Component folder is main folder for developing application.
in this folder we create folder for each component and that gives us power to have separate system for each component but with possibility to reuse everything.

<img width="162" alt="image" src="https://github.com/becarselman/ant-colony-pharaoh-frontend/assets/102960105/2c7fb667-6f4d-4222-8691-3ad3cd28e603">
* Now single component folder will look like this. We have View.js file as main file, component scss file Container file for importing stuff from state and actions and index.js file for connecting view and container.
<img width="159" alt="image" src="https://github.com/becarselman/ant-colony-pharaoh-frontend/assets/102960105/ae123170-2b60-4fa1-9e77-f1de068bd43c">
*In modules folder we put everything regarding state management for that component. For example actions for getChildren, setChildren... and sagas for that actions ....
*In component folder we will put helper components for this Children component (childrenItem, columns...)


If some component can be reusable in more screens we will create separate folder for them for example this Tables folder
<img width="257" alt="image" src="https://github.com/becarselman/ant-colony-pharaoh-frontend/assets/102960105/d5029b9f-f817-40d5-b238-61975957f49b">



Please use this structure in your project and your life will get easier when this app get bigger. @TarikVelic @ElvaRivle @meliha01

TarikVelic committed:
TariikMVP/FEATURE/ProjectsTable

--------------------------------------------------------------------
 MAIN
 * Pagination develop WIP
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,75 @@
+    <div className="table-container">
+      <Table dataSource={pageData} columns={columns} className="table" pagination={false} />
+
+      <div className="pagination-container">
```
You can move this whole part in separate component in Tables/components folder
You can call them something like noOfRows or something like that

becarselman commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,75 @@
+        <div className="pagination-info">
+          <span className="rows-text">Rows per page:</span>
+          <Select value={pageSize.toString()} onChange={handlePageSizeChange}>
+            {[1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map((value) => (
```
I think 1-10 is bad choice. I will go with 10 20 50. and also move this array outside of jsx ( connecting to previous comment you will move this outside of jsx in separate component noOfRows)

becarselman commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,75 @@
+          <span className="rows-text">Rows per page:</span>
+          <Select value={pageSize.toString()} onChange={handlePageSizeChange}>
+            {[1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map((value) => (
+              <Select.Option key={value} value={value.toString()}>
```
I see you use second time value.toString. Why you don't write initial value in array like string "10", "20" ...

becarselman commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,75 @@
+        <div className="pagination-right">
+        <Pagination {...paginationConfig} className="custom-pagination" showQuickJumper={false} />
+        <div className="next-last-buttons">
+          <button onClick={handleNextPage} disabled={currentPage === Math.ceil(totalItems / pageSize)}>
```
move this condition outside of jsx

becarselman commented on src/components/Table/CustomTable.scss:
```diff
@@ -0,0 +1,134 @@
+  display: flex;
+  align-items: center;
+
+  .ant-pagination-prev,
```
here you have example how to change antd arrows for prev and next page https://codesandbox.io/s/pagination-settings-antd-4-21-6-forked-0422vs?file=/demo.js

TarikVelic committed:
TariikMVP/FEATURE/ProjectsTable

--------------------------------------------------------------------
 MAIN
 * Review fixes
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on src/components/Table/CustomTable.scss:
```diff
@@ -0,0 +1,160 @@
+    margin-top: 16px;
+
+    .pagination-info {
+      @include gilroy-medium();
```
why you create mixin as function here.

becarselman commented on src/components/Table/CustomTable.scss:
```diff
@@ -0,0 +1,160 @@
+@import url('https://fonts.googleapis.com/css2?family=Inter:wght@400;500&display=swap');
```
you don't need to import font here if you are using one from mixin

becarselman commented on src/components/Table/CustomTable.scss:
```diff
@@ -0,0 +1,160 @@
+}
+
+.ant-pagination-next, 
+.ant-pagination-prev {
```
I don't like this solution

becarselman commented on src/components/Table/CustomTable.scss:
```diff
@@ -0,0 +1,160 @@
+  align-items: center;
+
+  .ant-pagination-item-active {
+    border-color: $gray-color !important;
```
try to avoid !important

becarselman commented on src/components/Table/Pagination.js:
```diff
@@ -0,0 +1,81 @@
+
+  const pageSizeSelect = (
+    <Select value={pageSize.toString()} onChange={handlePageSizeChange}>
+      {['5', '10', '20'].map((value) => (
```
One more time you ignore my comment. 1. move map outside of jsx
2. move array in separate variable. 3. use 5 10 and 20 entries is small amount of data, use 10 20 50.

becarselman commented on src/components/projects/components/data.js:
```diff
@@ -0,0 +1,288 @@
+export const tableData = [
+    {
+      key: '1',
```
you can use chatgtp to create this data. or you can create unique key by your self

TarikVelic committed:
TariikMVP/FEATURE/ProjectsTable

--------------------------------------------------------------------
 MAIN
 * review fixes
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
WIP Projects frontend

TarikVelic committed:
Project Frontend WIP

TarikVelic committed:
Added filter for projectStatus, WIP

TarikVelic committed:
Added search and filter functionality WIP Projects

TarikVelic committed:
Projects Frontend WIP

TarikVelic committed:
Small changes in styles. WIP

TarikVelic committed:
Created Saga, reducer, actions etc. WIP

AntMarkoLovric commented on package.json:
```diff
@@ -4,11 +4,14 @@
+    "antd": "^5.5.1",
     "axios": "^1.3.5",
     "cors": "^2.8.5",
+    "numeral": "^2.0.6",
```
Do we need to import this library or we can format it another way?

AntMarkoLovric commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,74 @@
+import React, { useState, useEffect } from 'react';
+import { Table, Spin } from 'antd';
```
Please format import as we discussed in the previous PR
-React import
-Third party import
-Local import
-Images
-SCSS files

AntMarkoLovric commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,74 @@
+    onPageChange(page);
+  };
+
+  useEffect(() => {
```
why do you use this in the use effect block, can we implement onClick function which will change value of the filter in state

AntMarkoLovric commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,74 @@
+  }, [data, selectedNavLabel]);
+
+  const handleSearch = (input) => {
+    onSearchChange(input);
```
i think it is better to save these state values in redux state instead of component state (useState). Then we can use it and access it throughout the whole project

AntMarkoLovric commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,74 @@
+    <div className="custom-table-container">
+      <Navbar
+        navLabels={navLabels} 
+        handlePageSelect={(page, label) => {
```
move inline functions from jsx

AntMarkoLovric commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,74 @@
+        }}
+      />
+      <div className="table-container">
+        {isLoading ? (
```
move ternary operator from jsx

AntMarkoLovric commented on src/components/Table/CustomTable.scss:
```diff
@@ -0,0 +1,151 @@
+
+    a {
+      color: $black-color;
+      background-color: #F5F6FB;
```
move this color into colors file or reuse it

AntMarkoLovric commented on src/components/Table/components/Navbar.js:
```diff
@@ -0,0 +1,31 @@
+  return (
+    <div className="navbar-container">
+      <div className="navbar-links">
+        {navLabels.map((label, index) => (
```
move map outside of jsx

AntMarkoLovric commented on src/components/Table/components/Navbar.js:
```diff
@@ -0,0 +1,31 @@
+            key={index}
+            label={label}
+            onClick={() => {
+              handlePageSelect(index + 1, label);
```
also move on click function outside of map, it will allow this on click function to be remembered, and it will be called when necessary, and not generated every time when you render navbar

AntMarkoLovric commented on src/components/dashboard/Dashboard.js:
when you finish testing and working on feature, remove changes on files that are not in your task (Dashboard, Sidebar, Routing etc.)

AntMarkoLovric commented on src/components/projects/View.js:
```diff
@@ -0,0 +1,53 @@
+const Projects = () => {
+  const dispatch = useDispatch();
+  const [page, setPage] = useState(1);
+  const [pageSize, setPageSize] = useState(10);
```
I think if you move state values in redux state from one of previous comments, you can access here and also you can use actions for that

AntMarkoLovric commented on src/components/projects/View.js:
```diff
@@ -0,0 +1,53 @@
+  const navLabels = ['All Projects', 'Active', 'Inactive', 'Completed'];
+
+  useEffect(() => {
+    dispatch(fetchProjectsRequest(page, pageSize, selectedProjectStatus, searchInput));
```
move this to container file and trigger actions like Selman suggested when we discussed this in the past

AntMarkoLovric commented on src/components/projects/View.js:
```diff
@@ -0,0 +1,53 @@
+    dispatch(fetchProjectsRequest(page, pageSize, selectedProjectStatus, searchInput));
+  }, [dispatch, page, pageSize, selectedProjectStatus, searchInput]);
+
+  const dataSource = useSelector((state) => state.projects.projects);
```
when you move actions, you can move selector also and remove useSelector from component

AntMarkoLovric commented on src/components/projects/components/data.js:
why didn't you remove this if you don't use it?

AntMarkoLovric commented on src/components/projects/index.js:
this file is used to make connection between container and view.js

AntMarkoLovric commented on src/index.scss:
why did you change .css to .scss

AntMarkoLovric commented on src/reducers/projectsReducer.js:
project reducer goes into projects/modules/reducer, and with that approach we fulfill fractal folder structure

AntMarkoLovric commented on src/sagas/projectsSaga.js:
also move this to projects/modules/saga

AntMarkoLovric commented on src/sagas/projectsSaga.js:
```diff
@@ -0,0 +1,68 @@
+
+
+function* fetchProjectsSaga(action) {
+  const { page, pageSize, selectedProjectStatus, searchInput } = action.payload;
```
without deconstructing this, you could pass an object to service and with that you can reduce number of parameters because best function is one without parameters

AntMarkoLovric commented on src/sagas/projectsSaga.js:
```diff
@@ -0,0 +1,68 @@
+    const projects = response.data.projects;
+    const total = response.data.count;
+
+    const formattedData = projects.map((project, index) => {
```
can we reduce complexity of this function using just one map, by removing map inside of a map

AntMarkoLovric commented on src/sagas/projectsSaga.js:
```diff
@@ -0,0 +1,68 @@
+    const total = response.data.count;
+
+    const formattedData = projects.map((project, index) => {
+      const developers = project.developers.map((developer) => developer.user);
```
you have two identical maps

AntMarkoLovric commented on src/sagas/projectsSaga.js:
```diff
@@ -0,0 +1,68 @@
+    yield put(fetchProjectsSuccess(formattedData, total));
+  } catch (error) {
+    yield put(fetchProjectsFailure(error));
+  } finally {
```
in "finally" block, we don't trigger fetch, we can, for example, turn off loader

TarikVelic committed:
WIP .data error Projects Frontend table

TarikVelic committed:
Review fixes update

TarikVelic committed:
TariikMVP/FEATURE/ProjectsTable

--------------------------------------------------------------------
 MAIN
 * Review comments fixed,
*  WIP
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,92 @@
+      className="table"
+      pagination={false}
+      bordered
+      title={() => tableHeader}
```
Why you create function here. I think you can just pass table Header.

becarselman commented on src/components/Table/components/Navbar.js:
```diff
@@ -0,0 +1,34 @@
+      key={index}
+      label={label}
+      onClick={() => handleClick(index, label)}
+      className={`${selectedPage === index + 1 ? 'active-link' : ''} ${
```
use classnames instead of this

becarselman commented on src/components/Table/modules/actions.js:
I can't see anywhere you use this actions. How I see you create this and leave old useState.

TarikVelic committed:
PUSH FOR DEMO

TarikVelic committed:
Scss update, WIP Projects Table

becarselman commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,92 @@
+  };
+
+  useEffect(() => {
+    const filteredData = data.filter((item) => {
```
you need to do this on backend. To be sure you get right data you will need to filter them on backend

becarselman commented on src/components/Table/components/Navbar.js:
```diff
@@ -0,0 +1,34 @@
+    <NavLink
+      key={index}
+      label={label}
+      onClick={() => handleClick(index, label)}
```
move this function outside of jsx and don't use inline function.

becarselman commented on src/components/Table/components/TableHeader.js:
```diff
@@ -0,0 +1,34 @@
+      </div>
+      <form
+        className="search-form"
+        onSubmit={(event) => {
```
move this on submit outside of jsx. this Whole block can be part of some function

becarselman commented on src/components/Table/components/TableHeader.js:
```diff
@@ -0,0 +1,34 @@
+          className="search-input"
+          placeholder="Search"
+          value={input}
+          onChange={(event) => setInput(event.target.value)}
```
same

becarselman commented on src/components/projects/components/columns.js:
```diff
@@ -0,0 +1,55 @@
+      render: (projectStatus) => (
+        <span>
+          <span
+           className={`status-dot ${projectStatus === 'Active' ? 
```
use classNames here  https://www.npmjs.com/package/classnames

TarikVelic committed:
Review Fixes change

TarikVelic committed:
Fixed Search Input

TarikVelic committed:
Added search input

TarikVelic committed:
Bug fixes WIP ProjectsTable

TarikVelic committed:
Loader update, WIP

becarselman commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,110 @@
+    setPageDataState(filteredData);
+  }, [data, filter]);
+
+  useEffect(() => {
```
Why you use this in use effect. Why you don't set onClick

becarselman commented on src/components/Table/components/TableHeader.js:
```diff
@@ -0,0 +1,54 @@
+
+  useEffect(() => {
+    setCountChanged(true);
+    const timeout = setTimeout(() => {
```
why you use timeout. Please find better solution.

becarselman commented on src/components/projects/View.js:
```diff
@@ -0,0 +1,64 @@
+
+  const fetchData = () => {
+    actions.fetchAllProjects(page, pageSize, selectedProjectStatus, searchInput);
+    dispatch(setPageData(null));
```
Don't use dispatch

becarselman commented on src/components/projects/View.js:
```diff
@@ -0,0 +1,64 @@
+  const handleNavSelect = (label) => {
+    setSelectedProjectStatus(label);
+    setPage(1);
+    dispatch(setPageData(null));
```
same

becarselman commented on src/components/sidebar/Sidebar.js:
```diff
@@ -1,51 +1,73 @@
+        to={option.path}
+        className={`sidebar-option ${isActive ? "active" : ""}`}
+        key={option.text}
+        onClick={() => handleClick(option.text)}
```
move out of jsx

becarselman commented on src/components/sidebar/Sidebar.scss:
```diff
@@ -1,22 +1,26 @@
+  height: 100%;
 
-  padding: 30px;
+  // max-width: 325px;
```
remove commented code

becarselman commented on src/components/sidebar/Sidebar.scss:
```diff
@@ -97,4 +100,53 @@
+  flex: 1;
+  background-color: #fff;
+  margin-left: 250px;
+  /* Dodati sledeći stil za poplavljivanje dužine */
```
use only english

becarselman commented on src/components/Table/components/Navbar.js:
```diff
@@ -0,0 +1,43 @@
+      <NavLink
+        key={index}
+        label={label}
+        onClick={handleNavLinkClick}
```
why you don't pass handle click instead of creating two functions

TarikVelic committed:
Review problems fixed

TarikVelic committed:
Deleting comments, small change

AntMarkoLovric commented on src/components/Table/CustomTable.js:
```diff
@@ -0,0 +1,110 @@
+  };
+
+  const handleClick = () => {
+    const filteredData = data.filter((item) => {
```
why you filter data if we get filtered data from backend

AntMarkoLovric commented on src/components/Table/components/Navbar.js:
```diff
@@ -0,0 +1,39 @@
+      <NavLink
+        key={index}
+        label={label}
+        onClick={() => handleClick(index, label)} 
```
move onClick outside of jsx

AntMarkoLovric commented on src/components/projects/View.js:
```diff
@@ -0,0 +1,61 @@
+  actions,
+}) => {
+  const [page, setPage] = useState(1);
+  const [pageSize, setPageSize] = useState(10);
```
This needs to be part of table component because it will be same for all tables

AntMarkoLovric commented on src/sagas/authSaga.js:
```diff
@@ -12,15 +12,27 @@ function* loginSaga(action) {
 }
 
-
+function* logoutSaga() {
```
this is not in your task, what did you do here?

AntMarkoLovric commented on src/service/projectsService.js:
```diff
@@ -0,0 +1,15 @@
+const URL = process.env.REACT_APP_API_URL;
+
+export const getAllProjects = ({ page, pageSize, selectedProjectStatus, searchInput }) => {
+  const statusQueryParam = selectedProjectStatus !== 'All Projects' ? selectedProjectStatus : '';
```
can we do this in saga?

AntMarkoLovric commented on src/service/projectsService.js:
```diff
@@ -0,0 +1,15 @@
+
+export const getAllProjects = ({ page, pageSize, selectedProjectStatus, searchInput }) => {
+  const statusQueryParam = selectedProjectStatus !== 'All Projects' ? selectedProjectStatus : '';
+  const searchQueryParam = searchInput ? `search[name]=${encodeURIComponent(searchInput)}` : '';
```
same as the comment above

AntMarkoLovric commented on src/components/Table/CustomTable.scss:
```diff
@@ -0,0 +1,212 @@
+    will-change: transform;
+    cursor: pointer;
+  
+    &:hover {
```
i know you have good intention with using on hover, but this doesn't look good

TarikVelic committed:
TariikMVP/FEATURE/ProjectsTable

--------------------------------------------------------------------
 MAIN
 * Review comments fixed
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
TariikMVP/FEATURE/ProjectsTable

--------------------------------------------------------------------
 MAIN
 * Fixed bug with Navbar
 * Loader centered
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on src/components/Table/CustomTable.scss:
```diff
@@ -0,0 +1,222 @@
+
+  .pagination-info {
+    @include gilroy-medium500;
+    color: #6C6D75;
```
move color into colors file

becarselman commented on src/components/Table/components/Navlink.js:
```diff
@@ -0,0 +1,16 @@
+import './Navlink.scss';
+
+const NavLink = ({ label, onClick, className }) => {
+  const handleClick = () => {
```
You don't need this function you can use onClick in jsx as well. this function just call function from props. Why you don't use directly

becarselman commented on src/components/Table/components/TableHeader.js:
```diff
@@ -0,0 +1,45 @@
+  return (
+    <div className="table-header">
+      <div className="header-content">
+        <h2 className="header-title">All Projects</h2>
```
Because we need to reuse this we can not use All Projects here. we need to get this title from props. And then it will be dynamic. If we go to employees page we want to se all employees instead of project. Am I right

becarselman commented on src/components/Table/components/TableHeader.scss:
```diff
@@ -0,0 +1,85 @@
+      align-items: center;
+      justify-content: center;
+      border-radius: 0.25rem;
+      padding: 0.625rem;
```
is this 10px?. This is so hard to read.

becarselman replied:
And also I think this is not to much important to be in relative units. We can fix this to suite design

becarselman commented on src/components/Table/components/TableHeader.scss:
```diff
@@ -0,0 +1,85 @@
+      justify-content: center;
+      border-radius: 0.25rem;
+      padding: 0.625rem;
+      font-size: 0.875rem;
```
14px?

becarselman commented on src/components/projects/View.js:
```diff
@@ -0,0 +1,55 @@
+  };
+
+  useEffect(() => {
+    fetchData(1, 10);
```
Leave this for now. But I think this will become really messy in future with all this numbers. My personal opinion is to reset page number and page size in redux state. But for now leave

becarselman commented on src/components/projects/modules/saga.js:
```diff
@@ -0,0 +1,68 @@
+
+function generateQueryParams(selectedProjectStatus, searchInput) {
+  const statusQueryParam = selectedProjectStatus !== 'All Projects' ? selectedProjectStatus : '';
+  const searchQueryParam = searchInput ? `search[name]=${encodeURIComponent(searchInput)}` : '';
```
Why you encode this. If user type name of project its visible already. He don't send anything private or important. he just type search

TarikVelic committed:
Review comments fixed
